api/v1alpha1/validations.go

```diff
@@ -0,0 +1,177 @@
+package v1alpha1
+
+import (
+	"fmt"
+
+	securityv1beta1 "istio.io/api/security/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	minValidPort int32 = 1024
+	maxValidPort int32 = 65535
+)
+
+func validateIstioMutualTLSMode(mode string) *field.Error {
+	if mode == "" {
+		return nil
+	}
+	if _, ok := securityv1beta1.PeerAuthentication_MutualTLS_Mode_value[mode]; ok {
+		return nil
+	}
+
+	var validModes []string
+	for v := range securityv1beta1.PeerAuthentication_MutualTLS_Mode_value {
+		validModes = append(validModes, v)
+	}
+
+	return field.Invalid(
+		field.NewPath("spec", "istioMutualTLSMode"),
+		mode,
+		fmt.Sprintf("mode must be one of the following: %v", validModes),
+	)
+}
+
+func validateWorkerReplicas(replicas *int32) *field.Error {
+	if replicas == nil || *replicas >= 0 {
+		return nil
+	}
+
+	return field.Invalid(
+		field.NewPath("spec", "worker", "replicas"),
+		replicas,
+		"should be greater than or equal to 0",
+	)
+}
+
+func validateWorkerResourceRequests(resources corev1.ResourceRequirements) field.ErrorList {
+	var errs field.ErrorList
+	fp := field.NewPath("spec", "worker", "resources", "requests")
+
+	if _, ok := resources.Requests[corev1.ResourceCPU]; !ok {
+		errs = append(errs, field.Required(fp.Child("cpu"), "is mandatory"))
+	}
+	if _, ok := resources.Requests[corev1.ResourceMemory]; !ok {
+		errs = append(errs, field.Required(fp.Child("memory"), "is mandatory"))
+	}
+
+	return errs
+}
+
+func validateImage(image *OCIImageDefinition) field.ErrorList {
+	var errs field.ErrorList
+	fp := field.NewPath("spec", "image")
+
+	if image.Repository == "" {
+		errs = append(errs, field.Required(fp.Child("repository"), "cannot be blank"))
+	}
+	if image.Tag == "" {
+		errs = append(errs, field.Required(fp.Child("tag"), "cannot be blank"))
+	}
+
+	return errs
+}
+
+func validateAutoscaler(as *Autoscaling) field.ErrorList {
+	if as == nil {
+		return nil
+	}
+
+	var errs field.ErrorList
+	fp := field.NewPath("spec", "autoscaling")
+
+	if as.MinReplicas != nil {
+		if *as.MinReplicas < 1 {
+			errs = append(errs, field.Invalid(
+				fp.Child("minReplicas"),
+				as.MinReplicas,
+				"must be greater than or equal to 1",
+			))
+		}
+
+		if *as.MinReplicas > as.MaxReplicas {
+			errs = append(errs, field.Invalid(
+				fp.Child("maxReplicas"),
+				as.MaxReplicas,
+				"cannot be less than spec.autoscaling.minReplicas",
+			))
+		}
+	}
+
+	if as.MaxReplicas < 1 {
+		errs = append(errs, field.Invalid(
+			fp.Child("maxReplicas"),
+			as.MaxReplicas,
+			"must be greater than or equal to 1",
+		))
+	}
+
+	if as.AverageCPUUtilization != nil && *as.AverageCPUUtilization <= 0 {
+		errs = append(errs, field.Invalid(
+			fp.Child("averageUtilization"),
+			as.AverageCPUUtilization,
+			"must be greater than 0",
+		))
+	}
```

Comment:
When autoscaling is enabled, `averageUtilization` should be an optional field because autoscaling may use memory instead of cpu utilization as the metric, or some other custom metric like gpu.  When neither cpu nor memory utilization is specified, HPA will default to 80% cpu utilization.

Btw, `averageUtilization` should be renamed to `averageCpuUtilization` or something similar.

Reply:
So, I'm happy to discuss a re-design of this feature addition but, just to be clear, this is exactly how it works for Ray and to my knowledge memory utilization and/or defaulting was never specified from the onset of this effort. Let's discuss it offline.